### ISSUE TITLE
Git assetstore

### ISSUE THREAD

**Issue by jeffbaumes** [girder/girder#1459]
@curtislisle had an idea for enabling a Git repository to be an assetstore or mount point in Girder. The rationale is that some information in Girder makes sense to track externally. There are currently some apps like Resonant Lab, Flow, maybe others, that have custom scripts that ingest data, analyses, etc. from Git repos into Girder items. This plugin would make this process a straightforward mount, and would allow the repo to be updated when corresponding data changes in Girder. The Git commit history would also be a great provenance tracker.

A question is how feasible it would it be to make a Git read/write assetstore? How much effort would that be?

**Comment by zachmullen** [comment]
This sounds like a filesystem assetstore with additional synchronization concerns, or is there more to it?

**Comment by manthey** [comment]
Based on my current work of using database-based assetstores, I think this
is completely feasible.  As with anything that should be synchronized, how
do we want Girder to check if the git repo has changed?  For instance, it a
git user makes a commit, Girder should update the files in some manner.

I could see the assetstore specifying the branch to use, too.

On Thu, Jul 14, 2016 at 11:24 AM, Jeffrey Baumes <notifications@github.com>
wrote:

> @curtislisle <https://github.com/curtislisle> had an idea for enabling a
> Git repository to be an assetstore or mount point in Girder. The rationale
> is that some information in Girder makes sense to track externally. There
> are currently some apps like Resonant Lab, Flow, maybe others, that have
> custom scripts that ingest data, analyses, etc. from Git repos into Girder
> items. This plugin would make this process a straightforward mount, and
> would allow the repo to be updated when corresponding data changes in
> Girder. The Git commit history would also be a great provenance tracker.
>
> A question is how feasible it would it be to make a Git read/write
> assetstore? How much effort would that be?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/girder/girder/issues/1459>, or mute the thread
> <https://github.com/notifications/unsubscribe/AIX_R5i8ucuxRyZavbVEy5-f3W3p8qsrks5qVlSegaJpZM4JMjGH>
> .
>



-- 
David Manthey
R&D Engineer
Kitware Inc.
(518) 881-4439

**Comment by zachmullen** [comment]
I agree, I think this is probably feasible, depending on the design. First and foremost, we need to define a mapping between git concepts and assetstore concepts. Assetstores essentially only facilitate uploading and downloading, whereas git has lots of additional concerns. Assetstores typically don't store history of a blob, for instance.

As far as writing, I think it would be fairly straightforward. On upload, we could construct a commit using the uploader's name and email and an auto-generated message. We can handle this for both new files and updating the contents of existing files entirely within the assetstore implementation.

**Comment by curtislisle** [comment]
It seems to me that checking for external updates could be an intentional Girder API call to access (git fetch) and merge when desired.  Then apps (like Flow) could use the girder api to check and merge changes on user request.   Also, I see and agree with Zach’s suggestion  that just came in while I was typing.

**Comment by jeffbaumes** [comment]
Thanks for the input. I think this proves feasibility enough to consider implementing. I'll work with @curtislisle and others to see how we could get this done.

**Comment by manthey** [comment]
I think this is a superset of issue #853.

**Comment by manthey** [comment]
Closing as outside of the main repo's scope.
